src/ngx_http_lua_args.c

```diff
@@ -42,8 +80,37 @@ ngx_http_lua_ngx_req_set_uri_args(lua_State *L)
 
     switch (lua_type(L, 1)) {
     case LUA_TNUMBER:
+        p = (u_char *) lua_tolstring(L, 1, &len);
+
+        args.data = ngx_palloc(r->pool, len);
+        if (args.data == NULL) {
+            return luaL_error(L, "no memory");
+        }
+
+        ngx_memcpy(args.data, p, len);
+
+        args.len = len;
+        break;
+
     case LUA_TSTRING:
         p = (u_char *) lua_tolstring(L, 1, &len);
+        if (ngx_http_lua_is_escaped_query_string(r, p, len, &b) != NGX_OK) {
+            buf_len = ngx_http_lua_escape_log(NULL, p, len);
+            buf = ngx_palloc(r->pool, buf_len + 1);
+            if (buf == NULL) {
+                return NGX_ERROR;
+            }
+
+            ngx_http_lua_escape_log(buf, p, len);
+            /* lua_pushfstring should be zero terminated */
+            buf[buf_len] = '\0';
+            msg = lua_pushfstring(L, "unescaped byte 0x%x in query string "
+                                  "\"%s\"", b, buf);
+
+            ngx_pfree(r->pool, buf);
+
+            return luaL_argerror(L, 1, msg);
```

Comment:
Sorry, after thinking more about this, I believe a better approach is to simply escape all these control and non-printable ASCII bytes automatically according to the URI escaping rules. This is also the way with minimal impact to the existing user code.